<a href="https://colab.research.google.com/github/Vatsalp599/Vatsalp599-DataScience-GenAI-submissions/blob/main/Vatsal_P_Mock_Test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

I asked AI "  want you to code and then when i run it i should get option of uploading the file and create dataframe"

In [ ]:
import pandas as pd
from google.colab import files
import io

print("Please upload your CSV file:")
uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))
  # Read the uploaded file into a pandas DataFrame
  df = pd.read_csv(io.StringIO(uploaded[fn].decode('utf-8')))
  print("\nDataFrame created successfully. Here are the first 5 rows:")
  display(df.head())

Please upload your CSV file:


Saving loan_data.csv to loan_data (2).csv
User uploaded file "loan_data (2).csv" with length 3454959 bytes

DataFrame created successfully. Here are the first 5 rows:


age  years_employed  annual_income  credit_score  credit_history_years  \
0   40            17.2          25579           692                   5.3   
1   33             7.3          43087           627                   3.5   
2   42             1.1          20840           689                   8.4   
3   53             0.5          29147           692                   9.8   
4   32            12.5          63657           630                   7.2   

   savings_assets  current_debt  defaults_on_file  delinquencies_last_2yrs  \
0             895         10820                 0                        0   
1             169         16550                 0                        1   
2              17          7852                 0                        0   
3            1480         11603                 0                        1   
4             209         12424                 0                        0   

   derogatory_marks  loan_amount  interest_rate  debt_to_income_ratio  \
0                 0          600          17.02                 0.423   
1                 0        53300          14.10                 0.384   
2                 0         2100          18.33                 0.377   
3                 0         2900          18.74                 0.398   
4                 0        99600          13.92                 0.195   

   loan_to_income_ratio  payment_to_income_ratio  loan_status  
0                 0.023                    0.008            1  
1                 1.237                    0.412            0  
2                 0.101                    0.034            1  
3                 0.099                    0.033            1  
4                 1.565                    0.522            1

it gave first 5 rows, which is enough for ensuring that data is correctly classified, it gave out all columns

We dont know if there are any empty cells in the data, so i asked AI " now in the data set check if there are any missing values"

In [ ]:
missing_values = df.isnull().sum()
print("Missing values per column:")
print(missing_values)

Missing values per column:
age                        0
years_employed             0
annual_income              0
credit_score               0
credit_history_years       0
savings_assets             0
current_debt               0
defaults_on_file           0
delinquencies_last_2yrs    0
derogatory_marks           0
loan_amount                0
interest_rate              0
debt_to_income_ratio       0
loan_to_income_ratio       0
payment_to_income_ratio    0
loan_status                0
dtype: int64


As in the output we can see, there are no missing values in any column

Now i informed AI to ensure and keep in its memory that Y value is loan status and rest are X values by saying " remember that Y value = loan status and rest are X values"

In [ ]:
y = df['loan_status']
X = df.drop('loan_status', axis=1)

print("Shape of X (features):")
print(X.shape)
print("Shape of y (target):")
print(y.shape)
print("\nFirst 5 rows of X:")
display(X.head())
print("\nFirst 5 values of y:")
display(y.head())

Shape of X (features):
(50000, 15)
Shape of y (target):
(50000,)

First 5 rows of X:


age  years_employed  annual_income  credit_score  credit_history_years  \
0   40            17.2          25579           692                   5.3   
1   33             7.3          43087           627                   3.5   
2   42             1.1          20840           689                   8.4   
3   53             0.5          29147           692                   9.8   
4   32            12.5          63657           630                   7.2   

   savings_assets  current_debt  defaults_on_file  delinquencies_last_2yrs  \
0             895         10820                 0                        0   
1             169         16550                 0                        1   
2              17          7852                 0                        0   
3            1480         11603                 0                        1   
4             209         12424                 0                        0   

   derogatory_marks  loan_amount  interest_rate  debt_to_income_ratio  \
0                 0          600          17.02                 0.423   
1                 0        53300          14.10                 0.384   
2                 0         2100          18.33                 0.377   
3                 0         2900          18.74                 0.398   
4                 0        99600          13.92                 0.195   

   loan_to_income_ratio  payment_to_income_ratio  
0                 0.023                    0.008  
1                 1.237                    0.412  
2                 0.101                    0.034  
3                 0.099                    0.033  
4                 1.565                    0.522


First 5 values of y:


0    1
1    0
2    1
3    1
4    1
Name: loan_status, dtype: int64

As one can see, the AI wrote code which gave output by segreggating the X values and Y value and giving first 5 row values of it

Now as i want to perform modelling by machine learning, i asked AI " divide dataset into training and testing split with 75% in training and 25% in testing and also stratify= Y value"

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42, stratify=y)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (37500, 15)
Shape of X_test: (12500, 15)
Shape of y_train: (37500,)
Shape of y_test: (12500,)


AI coded and separated the data into training and testing split in 75% training and 25% testing, it has given us the shape of X train, X test, Y train, Y test along with number of rows and columns with 37500 rows being in train and 12500 being in test which is exactly 75% and 25% of total rows in inital dataset

Before carrying out modelling with help of machine learning, i asked AI to scale the data with help of min-max scaling to ensure even weightage of all the X values, " scale the dataset with the help of min-max scaler

In [ ]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Convert scaled arrays back to DataFrames for easier viewing and further processing (optional)
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns, index=X_train.index)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns, index=X_test.index)

print("First 5 rows of scaled X_train:")
display(X_train_scaled.head())
print("\nFirst 5 rows of scaled X_test:")
display(X_test_scaled.head())

First 5 rows of scaled X_train:


age  years_employed  annual_income  credit_score  \
27643  0.153846        0.082707       0.269970      0.468127   
46364  0.500000        0.047619       0.000000      0.723108   
48000  0.461538        0.137845       0.249451      0.721116   
1671   0.230769        0.152882       0.268204      0.657371   
20729  0.615385        0.150376       0.251783      0.711155   

       credit_history_years  savings_assets  current_debt  defaults_on_file  \
27643              0.060000        0.010547      0.257387               0.0   
46364              0.470000        0.007457      0.039506               0.0   
48000              0.023333        0.000107      0.144664               0.0   
1671               0.130000        0.014100      0.093521               0.0   
20729              0.790000        0.016077      0.266243               0.0   

       delinquencies_last_2yrs  derogatory_marks  loan_amount  interest_rate  \
27643                      0.0              0.25     0.714573       0.584118   
46364                      0.0              0.00     0.010050       0.581765   
48000                      0.0              0.00     0.698492       0.564706   
1671                       0.0              0.00     0.510553       0.361176   
20729                      0.0              0.00     0.698492       0.620000   

       debt_to_income_ratio  loan_to_income_ratio  payment_to_income_ratio  
27643              0.670426              0.454089                 0.453313  
46364              0.542607              0.046162                 0.045181  
48000              0.401003              0.473156                 0.472892  
1671               0.244361              0.325640                 0.325301  
20729              0.733083              0.469644                 0.469880


First 5 rows of scaled X_test:


age  years_employed  annual_income  credit_score  \
8543   0.307692        0.177945       0.060574      0.535857   
47570  0.038462        0.025063       0.392834      0.472112   
5406   0.000000        0.025063       0.024523      0.760956   
37876  0.596154        0.022556       0.000000      0.488048   
6485   0.884615        0.010025       0.167519      0.555777   

       credit_history_years  savings_assets  current_debt  defaults_on_file  \
8543               0.340000        0.000853      0.023912               0.0   
47570              0.033333        0.000617      0.119074               0.0   
5406               0.050000        0.000173      0.045973               0.0   
37876              0.996667        0.044730      0.008471               0.0   
6485               0.153333        0.005907      0.081418               0.0   

       delinquencies_last_2yrs  derogatory_marks  loan_amount  interest_rate  \
8543                  0.222222               0.0     0.493467       0.443529   
47570                 0.222222               0.0     0.772864       0.568824   
5406                  0.000000               0.0     0.128643       0.612353   
37876                 0.000000               0.0     0.248241       0.548824   
6485                  0.000000               0.0     0.210050       0.404706   

       debt_to_income_ratio  loan_to_income_ratio  payment_to_income_ratio  
8543               0.167920              0.847466                 0.847892  
47570              0.225564              0.357752                 0.356928  
5406               0.454887              0.317612                 0.317771  
37876              0.119048              0.838936                 0.838855  
6485               0.305764              0.193678                 0.192771

The code by AI has scaled the X train and X test dataset with the help of Min-Max scaler and i can see all the values are in between of 0 and 1 which is what we were expecting

Now i asked AI to perform logistic regression to predict whether loan will be approved or rejected and then get us scores of the model for analysis purpose, " here generate the logistic regression to predict if the loan is accepted or rejected and also give the scores of the model for analysis purpose"

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Initialize the Logistic Regression model
log_reg_model = LogisticRegression(random_state=42)

# Train the model on the scaled training data
log_reg_model.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred = log_reg_model.predict(X_test_scaled)

# Evaluate the Logistic Regression model's performance
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Logistic Regression Model Performance (Full Model):")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-Score: {f1:.4f}")

Logistic Regression Model Performance (Full Model):
Accuracy: 0.8379
Precision: 0.8446
Recall: 0.8647
F1-Score: 0.8545


The logistic regression model achieves an accuracy of 83.79%, indicating overall good performance. For the loan approved class, the precision is 84.46%, meaning that when the model predicts approval, it is correct about 84% of the time. The recall is 86.47%, showing that the model successfully identifies about 86% of all actual approved loans. This suggests the model is slightly better at capturing approved loans than avoiding false approvals, but loans for ineligible candidates getting not approved is the priority so we need less false positives and hence better precision score

For better precision, we need less false positives and model to have better score for precision, hence i asked AI to increase score of precision by using hyperparameters in the model and using gridsearch, " tune a model with help of hyperparameters and with help of grid search to have high precision and then run it"



In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'class_weight': [None, 'balanced']
}

print("Libraries imported and parameter grid defined.")

Libraries imported and parameter grid defined.


In [ ]:
logistic_model = LogisticRegression(solver='liblinear', random_state=42)
grid_search = GridSearchCV(estimator=logistic_model, param_grid=param_grid, scoring='precision', cv=5, verbose=1, n_jobs=-1)
grid_search.fit(X_train_scaled, y_train)

print(f"Best parameters: {grid_search.best_params_}")
print(f"Best precision score: {grid_search.best_score_}")

Fitting 5 folds for each of 12 candidates, totalling 60 fits
Best parameters: {'C': 100, 'class_weight': 'balanced'}
Best precision score: 0.8589663932501571


As we can see the AI coded for gridsearchCV, got best hyperparameter and tuned model with precision of 0.8594 and then ran the tuned model to get scores,
The logistic regression model achieves an overall accuracy of 83.61%. For loan approval predictions, the precision is 85.94%, meaning that when the model predicts approval, it is correct about 86% of the time. This high precision aligns with the objective of minimizing approvals for ineligible applicants.

Now i Asked AI to generate a model using decision trees" i want you to generate decision tree now to predict whether loan gets approved or rejected and give scores than for analysis purpose"

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Initialize the Decision Tree Classifier
decision_tree_model = DecisionTreeClassifier(random_state=42)

# Train the model on the scaled training data
decision_tree_model.fit(X_train_scaled, y_train)

# Make predictions on the scaled test data
y_pred_dt = decision_tree_model.predict(X_test_scaled)

# Evaluate the Decision Tree model's performance
accuracy_dt = accuracy_score(y_test, y_pred_dt)
precision_dt = precision_score(y_test, y_pred_dt)
recall_dt = recall_score(y_test, y_pred_dt)
f1_dt = f1_score(y_test, y_pred_dt)

print(f"Decision Tree Model Performance:")
print(f"Accuracy: {accuracy_dt:.4f}")
print(f"Precision: {precision_dt:.4f}")
print(f"Recall: {recall_dt:.4f}")
print(f"F1-Score: {f1_dt:.4f}")

Decision Tree Model Performance:
Accuracy: 0.8322
Precision: 0.8483
Recall: 0.8465
F1-Score: 0.8474


The Decision Tree model achieves an accuracy of 83.22%, indicating solid overall performance. For loan approval predictions, the precision is 84.83%, meaning that when the model predicts approval, it is correct about 85% of the time. Compared to logistic regression, this model offers similar accuracy but slightly lower precision, which may be relevant if minimizing false approvals is the priority

i asked AI to use hyperparameters to train decision tree model in order to increase precision score. "  i want you to train decision tree model with hyperparameters and tuning with gridsearchCV for the best precision score"

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

param_grid = {
    'max_depth': [None, 5, 10, 20],
    'min_samples_split': [2, 10, 20],
    'min_samples_leaf': [1, 5, 10],
    'criterion': ['gini', 'entropy']
}

print("Libraries imported and parameter grid for Decision Tree defined.")

Libraries imported and parameter grid for Decision Tree defined.


In [ ]:
dt_model = DecisionTreeClassifier(random_state=42)
grid_search_dt = GridSearchCV(estimator=dt_model, param_grid=param_grid, scoring='precision', cv=5, verbose=1, n_jobs=-1)
grid_search_dt.fit(X_train_scaled, y_train)

print(f"Best parameters for Decision Tree: {grid_search_dt.best_params_}")
print(f"Best precision score for Decision Tree: {grid_search_dt.best_score_}")

Fitting 5 folds for each of 72 candidates, totalling 360 fits
Best parameters for Decision Tree: {'criterion': 'gini', 'max_depth': None, 'min_samples_leaf': 10, 'min_samples_split': 2}
Best precision score for Decision Tree: 0.872913547908777


In [ ]:
best_dt_model = grid_search_dt.best_estimator_
y_pred_dt_tuned = best_dt_model.predict(X_test_scaled)

accuracy_dt_tuned = accuracy_score(y_test, y_pred_dt_tuned)
precision_dt_tuned = precision_score(y_test, y_pred_dt_tuned)
recall_dt_tuned = recall_score(y_test, y_pred_dt_tuned)
f1_dt_tuned = f1_score(y_test, y_pred_dt_tuned)

print("Tuned Decision Tree Model Performance:")
print(f"Accuracy: {accuracy_dt_tuned:.4f}")
print(f"Precision: {precision_dt_tuned:.4f}")
print(f"Recall: {recall_dt_tuned:.4f}")
print(f"F1-Score: {f1_dt_tuned:.4f}")

Tuned Decision Tree Model Performance:
Accuracy: 0.8537
Precision: 0.8699
Recall: 0.8632
F1-Score: 0.8666


After hyperparameter tuning, the Decision Tree model shows improved performance with an accuracy of 85.37%. Precision is 86.99%, meaning that when the model predicts a loan approval, it is correct nearly 87% of the time, which aligns well with the goal of minimizing approvals for ineligible applicants. Recall is 86.32%, indicating that the model successfully identifies most eligible applicants. Compared to the previous version, tuning has significantly enhanced precision and overall performance, making this model more suitable for risk-sensitive loan approval decisions.